crates/task-impls/src/consensus/mod.rs

```diff
@@ -616,9 +622,31 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> ConsensusTaskState<TYPES, I>
                         *disperse.data.view_number,
                     ))
                     .await;
-                if self.vote_if_able(&event_stream).await {
-                    self.current_proposal = None;
-                }
+                let Some(proposal) = self.current_proposal.clone() else {
+                    return;
+                };
+                let upgrade = self.decided_upgrade_cert.clone();
+                let pub_key = self.public_key.clone();
+                let priv_key = self.private_key.clone();
+                let consensus = Arc::clone(&self.consensus);
+                let storage = Arc::clone(&self.storage);
+                let quorum_mem = Arc::clone(&self.quorum_membership);
+                let committee_mem = Arc::clone(&self.committee_membership);
+                let instance_state = Arc::clone(&self.instance_state);
+                let handle = async_spawn(async move {
```

Comment:
Looks like this exact block is copied a couple of times, is it worth perhaps defining a helper function for it?